substrate/client/rpc-spec-v2/src/transaction/transaction_broadcast.rs

```diff
@@ -0,0 +1,217 @@
+	}
+
+	/// Generate an unique operation ID for the `transaction_broadcast` RPC method.
+	pub fn generate_unique_id(&self) -> String {
+		let generate_operation_id = || {
+			// The lenght of the operation ID.
+			const OPERATION_ID_LEN: usize = 16;
+
+			rand::thread_rng()
+				.sample_iter(Alphanumeric)
+				.take(OPERATION_ID_LEN)
+				.map(char::from)
+				.collect::<String>()
+		};
+
+		let mut id = generate_operation_id();
+
+		let broadcast_ids = self.broadcast_ids.read();
+
+		while broadcast_ids.contains_key(&id) {
+			id = generate_operation_id();
+		}
+
+		id
+	}
+}
+
+/// Currently we treat all RPC transactions as externals.
+///
+/// Possibly in the future we could allow opt-in for special treatment
+/// of such transactions, so that the block authors can inject
+/// some unique transactions via RPC and have them included in the pool.
+const TX_SOURCE: TransactionSource = TransactionSource::External;
+
+#[async_trait]
+impl<Pool, Client> TransactionBroadcastApiServer<BlockHash<Pool>>
+	for TransactionBroadcast<Pool, Client>
+where
+	Pool: TransactionPool + Sync + Send + 'static,
+	Pool::Hash: Unpin,
+	<Pool::Block as BlockT>::Hash: Unpin,
+	Client: HeaderBackend<Pool::Block> + BlockchainEvents<Pool::Block> + Send + Sync + 'static,
+{
+	fn broadcast(&self, bytes: Bytes) -> RpcResult<Option<String>> {
+		let pool = self.pool.clone();
+
+		// The unique ID of this operation.
+		let id = self.generate_unique_id();
+
+		let mut best_block_import_stream =
+			Box::pin(self.client.import_notification_stream().filter_map(
+				|notification| async move { notification.is_new_best.then_some(notification.hash) },
+			));
+
+		let broadcast_transaction_fut = async move {
+			// There is nothing we could do with an extrinsic of invalid format.
+			let Ok(decoded_extrinsic) = TransactionFor::<Pool>::decode(&mut &bytes[..]) else {
+				return
+			};
+
+			// Flag to determine if the we should broadcast the transaction again.
+			let mut is_done = false;
+
+			while !is_done {
+				// Wait for the next block to become available.
+				let Some(mut best_block_hash) = best_block_import_stream.next().await else {
+					return
+				};
+				// We are effectively polling the stream for the last available item at this time.
+				// The `now_or_never` returns `None` if the stream is `Pending`.
+				//
+				// If the stream contains `Hash0x1 Hash0x2 Hash0x3 Hash0x4`, we want only `Hash0x4`.
+				while let Some(next) = best_block_import_stream.next().now_or_never() {
+					let Some(next) = next else {
+						// Nothing to do if the best block stream terminated.
+						return
+					};
+					best_block_hash = next;
+				}
+
+				let submit =
+					pool.submit_and_watch(best_block_hash, TX_SOURCE, decoded_extrinsic.clone());
+
+				// The transaction was not included to the pool, because it is invalid.
+				// However an invalid transaction can become valid at a later time.
+				let Ok(mut stream) = submit.await else { return };
```

Comment:
Ahh yea, this is wrong. Must be a left-over comment, will remove it : D

Reply:
I've had another look over this while writing some extra tests.

It seems indeed that the tx did not enter the pool because of an error.

There's only one that we might need to care about: 

https://github.com/paritytech/polkadot-sdk/blob/6ea472ad5a68018dc054ec0a5d866fc481d077e4/substrate/client/transaction-pool/api/src/error.rs#L45-L46

However, I still think we should return here and do nothing the tx further. If I'm looking at the right place in the code, the default time for marking a tx as "banned" is 30 mins:

https://github.com/paritytech/polkadot-sdk/blob/6ea472ad5a68018dc054ec0a5d866fc481d077e4/substrate/client/transaction-pool/src/graph/rotator.rs#L49

It might turn out to be beneficial to add to the spec a soft limit "tx broadcast ensures the tx is broadcasted for at least a number of blocks (16) if the server chooses to validate the tx" -- then we could safely return from here; and when we drop this future we remove `BroadcastState` from internal tracking